### PR TITLE
GH-34543: [CI] Self-hosted ARM workflows improvements

### LINKED ISSUE #34543: [CI] Improve ARM64 GitHub actions jobs that were migrated from Travis
### Describe the enhancement requested

There were some comments after the PR for migrating the Travis ARM jobs to GitHub actions was merged. I create a new issue to resolve those comments because the PR has become bigger than originally expected and I consider is not MINOR anymore.
The original PR with the comments can be found here:
https://github.com/apache/arrow/pull/34482

### Component(s)

Continuous Integration

### PR DESCRIPTION
**Pull request by raulcd**
### Rationale for this change

Fixes some post-merge review comments

### What changes are included in this PR?

Comments from original: https://github.com/apache/arrow/pull/34482

### Are these changes tested?

On CI

* Closes: #34543

### CODE REVIEW

**Comment by raulcd** [comment]
The job failure is not related to the PR

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -42,20 +42,31 @@ permissions:
...
-          - go: 1.18
+          - arch-label: AMD64
+            arch: amd64
+            go: 1.18
+            runs-on: ubuntu-latest
+            staticcheck: v0.3.3
+          - arch-label: ARM64
+            arch: arm64v8
+            go: 1.17
             staticcheck: v0.3.3
```
Comment (kou) [inline .github/workflows/go.yml:66]: Could you use `v0.2.2` for Go 1.17 like the existing 1.17 job?

```suggestion
            staticcheck: v0.2.2
```

Could you also remove needless code?

```diff
diff --git a/ci/scripts/go_build.sh b/ci/scripts/go_build.sh
index c113bbd320..3c8cc0f4ee 100755
--- a/ci/scripts/go_build.sh
+++ b/ci/scripts/go_build.sh
@@ -20,13 +20,6 @@
 set -ex
 
 source_dir=${1}/go
-ARCH=`uname -m`
-
-# Arm64 CI is triggered by travis and run in arm64v8/golang:1.17-bullseye
-if [ "aarch64" == "$ARCH" ]; then
-# Install `staticcheck`
-  GO111MODULE=on go install honnef.co/go/tools/cmd/staticcheck@v0.2.2
-fi
 
 pushd ${source_dir}/arrow
 
```

File: .github/workflows/cpp.yml [hunk 2]
```diff
@@ -401,15 +401,14 @@ jobs:
           PATH="$(cygpath --unix ${PYTHON_BIN_DIR}):${PATH}"
           ci/scripts/cpp_test.sh "$(pwd)" "$(pwd)/build"
 
-  linux-arm:
+  docker-arm:
```
Comment (kou) [inline .github/workflows/cpp.yml:404]: Can we unify this to the existing `docker` job like Go's one?

File: .github/workflows/cpp.yml [hunk 3]
```diff
@@ -55,22 +55,39 @@ env:
...
     strategy:
       fail-fast: false
       matrix:
-        image:
-          - conda-cpp
-          - ubuntu-cpp-sanitizer
         include:
-          - image: conda-cpp
+          - arch: amd64
+            clang_tools: "14"
```
Comment (kou) [inline .github/workflows/cpp.yml:66]: Could you use `-` not `_` for word separator for `matrix` parameter names like `ruins-on`?

```suggestion
            clang-tools: "14"
```

Comment (raulcd) [reply .github/workflows/cpp.yml]: sure, I've done it. Just so I know in the future, is there a reason for it I am not aware of?

Comment (kou) [reply .github/workflows/cpp.yml]: Because it's GitHub Actions' convention.

Comment (raulcd) [reply .github/workflows/cpp.yml]: ok, that makes sense. I found some `_` sometimes but in general they seem to use `-`. There is even a question opened on if the convention exists: https://github.com/community/community/discussions/39547
I'll have it in mind in the future so at least we have a convention. Thanks @kou

**Review by kou — APPROVED** [review]
+1

**Pull request merged**